### PR TITLE
JIT: fix double reporting of some GC frame slots

### LINKED ISSUE #71005: [libraries-pgo] arm64 failure in System.Tests.DoubleTests.ParsePatterns
This test has been failing off and on for a while with random PGO (and possibly other modes)

One common failure mode is
```
set COMPlus_TieredCompilation=1
set COMPlus_ReadyToRun=0
set COMPlus_TC_QuickJitForLoops=1
set COMPlus_TieredPGO=1
set COMPlus_JitRandomEdgeCounts=1

    System.Tests.DoubleTests.ParsePatterns [FAIL]
      Assert.Equal() Failure
                      ↓ (pos 6)
      Expected: 00000062
      Actual:   00000000
                      ↑ (pos 6)
      Stack Trace:
        C:\repos\runtime0\src\libraries\System.Runtime\tests\System\DoubleTests.cs(433,0): at System.Tests.DoubleTests.ParsePatterns()
           at InvokeStub_DoubleTests.ParsePatterns(Object, Object, IntPtr*)
           at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```
Though I have also seen
```
  Starting:    System.Runtime.Tests (parallel test collections = off, max threads = 8)
    System.Tests.DoubleTests.ParsePatterns [FAIL]
      System.AccessViolationException : Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
      Stack Trace:
        C:\repos\runtime0\src\libraries\System.Private.CoreLib\src\System\BitConverter.cs(678,0): at System.BitConverter.<>c.<ToString>b__43_0(Span`1 dst, ValueTuple`3 state)
        C:\repos\runtime0\src\libraries\System.Runtime\tests\System\DoubleTests.cs(433,0): at System.Tests.DoubleTests.ParsePatterns()
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
        C:\repos\runtime0\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodInvoker.cs(69,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```
and gc heap corruption.

Repros fairly readily with just this test enabled (add `-parallel none -method System.Tests.DoubleTests.ParsePatterns` to xunit args).

Binary search via `COMPlus_JitEnablePgoRange` indicates the problematic method is
```
Compiling 7228 System.Tests.DoubleTests::ParsePatterns, IL size = 287, hash=0xb7003f40 Tier1-OSR @0x102
```

### PR DESCRIPTION
**Pull request by AndyAyersMS**
If there is a gc struct local that is dependently promoted, the struct
local may be untracked while the promoted gc fields of the struct are
tracked.

If so, the jit will double report the stack offset for the gc field,
first as an untracked slot, and then as a tracked slot.

Detect this case and report the slot as tracked only.

Closes #71005.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib PTAL

Not sure yet why this hasn't bitten us before, or (perhaps) only seems to have caused arm64 OSR failures.

**Comment by AndyAyersMS** [comment]
This seems to be the wrong approach. During `genFnProlog` we scan for on-frame locals that will require tracked reporting, and explicitly exclude dependently promoted structs. However, we keep track of these locals by a pair of frame offsets.

So evidently something goes wrong here (perhaps just with OSR) and we get this range wrong, or are not properly ordering the frame so that we're intermingling locals improperly or something. Continuing to investigate.

**Comment by SingleAccretion** [comment]
@AndyAyersMS FWIW, I investigated this earlier today, perhaps my conclusions could be of help.

<details>
<summary>Details (note these are purely from looking at the code and may be partially or completely incorrect.)</summary>

Ordinarily the frame allocation is done s.t. the "tracked GC refs on stack" are all contained in one region:
```
If we're supposed to track lifetimes of pointer temps, we'll
assign frame offsets in the following order:

    non-pointer local variables (also untracked pointer variables)
        pointer local variables
        pointer temps
    non-pointer temps
```
The way reporting of "tracked GC refs on stack" is done, codegen calculates the low/high offsets within which they reside:

https://github.com/dotnet/runtime/blob/6bc83f1b4e1f6f03cfd4a4548ac299364313b290/src/coreclr/jit/codegencommon.cpp#L5634-L5654

So, ordinarily, we would not have any dependently promoted fields in this range (since they'd be in the "untracked" group).

Presumably, OSR violates this by virtue of having "fixed" offsets for some locals.

When the emitter looks at which of these slots actually need to be reported here:

https://github.com/dotnet/runtime/blob/6bc83f1b4e1f6f03cfd4a4548ac299364313b290/src/coreclr/jit/emit.cpp#L6543-L6555

It calls `lvaIsGCTracked` that will always return `false` for dependently promoted fields on x64, but not ARM64 (note `emitContTrkPtrLcls` is always `false` on both targets).
</details>

**Comment by AndyAyersMS** [comment]
@SingleAccretion thanks, I have been looking at exactly this area.

Seems like there are two options:
(1) change filtering I'm doing now to use `lvaIsGCTracked` to filter out some cases where despite being a tracked GC ref, it is reported as untracked;
(2) change the call to `emitEndCodeGen` so that for OSR we don't claim to have a contiguous range of tracked stack locals (which ends up calling into `lvaIsGCTracked` under the covers).

**Comment by SingleAccretion** [comment]
Judging by the code which binds the offset range for "tracked on-stack GC refs", it seems to me that "dependently promoted fields" cannot really be tracked, otherwise we would need to include them into the range.

**Comment by AndyAyersMS** [comment]
For OSR the tracked range may end up including some trackable dependent promoted locals. They won't ever inspire or extend the range, but they may fall within the range set by other on frame gc vars.

I updated the code to detect this case and report as tracked, by querying the emitter to see if the offset falls within the tracked range. I suppose there's still a concern that we might not properly track the lifetimes (since say they can have aliased reads/writes via the parent) in which case we'd need to find a way to exclude them from tracked reporting somehow.

**Comment by AndyAyersMS** [comment]
Was hoping to scope this down to just an OSR change, but am seeing some non-OSR hits for this new check:
```
;; arm64 windows crossgen of SPC
;; System.IO.Enumeration.FileSystemEnumerator`1:InternalDispose(bool):this

1 tracked GC refs are at stack offsets  0010 ...  0018

;  V05 tmp1         [V05    ] (  2,  0   )  struct (24) [fp+18H]   do-not-enreg[HS] must-init hidden-struct-arg "struct address for call/obj"
;* V06 tmp2         [V06    ] (  0,  0   )     ref  ->  zero-ref    class-hnd "Inlining Arg"
;  V07 tmp3         [V07,T04] (  3,  0   )    long  ->   x0         "Inlining Arg"
;* V08 tmp4         [V08    ] (  0,  0   )    long  ->  zero-ref    "Inline stloc first use temp"
;  V09 tmp5         [V09,T06] (  1,  0   )     ref  ->  [fp+18H]   do-not-enreg[H] hidden-struct-arg V05.Item2(offs=0x00) P-DEP "field V05.Item2 (fldOffset=0x0)"
;  V10 tmp6         [V10,T05] (  2,  0   )    long  ->  [fp+20H]   do-not-enreg[H] hidden-struct-arg V05.Item1(offs=0x08) P-DEP "field V05.Item1 (fldOffset=0x8)"
;  V11 tmp7         [V11,T07] (  1,  0   )     int  ->  [fp+28H]   do-not-enreg[H] hidden-struct-arg V05.Item3(offs=0x10) P-DEP "field V05.Item3 (fldOffset=0x10)"
;  V12 PSPSym       [V12,T02] (  1,  1   )    long  ->  [fp+38H]   do-not-enreg[V] "PSPSym"

Untracked GC struct slot V05+0 (P-DEP promoted V09) is at frame offset 24 within tracked ref range.
```
In this case `V09` only appears as a child of its parent and never becomes live. Looks like the parent is an ignored gc struct return value.

Or maybe I need to exclude the upper end of the offset range. That's probably it.

**Comment by AndyAyersMS** [comment]
Looks like may be arm with profiler callbacks ends up tripping this; likely need to exempt "prespilled" cases too.

**Comment by AndyAyersMS** [comment]
PTAL @dotnet/jit-contrib 

Libraries-pgo has some known failures so I didn't expect it to run cleanly.

**Comment by SingleAccretion** [comment]
> I suppose there's still a concern that we might not properly track these lifetimes

Yes... I think some unrolled varieties of block copies, especially those using SIMD registers, do not properly report GC-ness for the destination stack slots. Likewise for our special calls with tracked "return buffers".

**Comment by AndyAyersMS** [comment]
> > I suppose there's still a concern that we might not properly track these lifetimes
> 
> Yes... I think some unrolled varieties of block copies, especially those using SIMD registers, do not properly report GC-ness for the destination stack slots. Likewise for our special calls with tracked "return buffers".

Ok, I've switched things around to always report as untracked.

File: src/coreclr/jit/compiler.hpp [hunk 1]
```diff
@@ -4021,7 +4021,7 @@ inline bool Compiler::lvaIsGCTracked(const LclVarDsc* varDsc)
 #ifdef TARGET_AMD64
         return !isStackParam && !lvaIsFieldOfDependentlyPromotedStruct(varDsc);
 #else  // !TARGET_AMD64
-        return !isStackParam;
+        return !isStackParam && !lvaIsFieldOfDependentlyPromotedStruct(varDsc);
 #endif // !TARGET_AMD64
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.hpp:4025]: Unnecessary ifdef now?

**Review by SingleAccretion — APPROVED** [review]
LGTM modulo nits.

**Review by BruceForstall — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
Was hoping this plus other recent PRs would have cleaned up libraries-pgo, but there is still more going on there. Going to see if the new slew of arm64 failures repro locally (looks like a machine issue but I can't tell for sure).

**Comment by AndyAyersMS** [comment]
Kicked off a new libraries-pgo baseline to see what that looks like.

**Comment by AndyAyersMS** [comment]
libraries-pgo picture is still a bit confusing but I think this change is good.

**Pull request merged**